### PR TITLE
Correctly propagate Maybe handlers. Closes #102 and #119.

### LINKED ISSUE #102: Nullable handler doesn't support objects
Probably.

Code is:

```haskell
instance forall m hg. (HasResolver m hg, Functor m, ToValue (Maybe hg)) => HasResolver m (Maybe hg) where
  type Handler m (Maybe hg) = m (Maybe hg)
  resolve handler _ =  map (ok . toValue) handler
```

Note that it discards the selection set. That can't be good.

### PR DESCRIPTION
**Pull request by teh**
I'm not happy with this solution because of the extra monad level
introduced.

    pure (Just (pure value))

is not a good way to return a Maybe. Not sure how to fix better ATM.

### CODE REVIEW

**Comment by jml** [comment]
I don't like the extra monad level either. I'm not sure how to fix and I don't think I fully understand what's going on. OK to hold off a little to give us both time to think of something better?

**Comment by teh** [comment]
The extra level happens because every "type translation" step in form of an `instance HasResolver` introduces a new monad level ATM.
```
  type Handler m (Maybe hg) = m (Maybe (Handler m hg))
                                        ^^^^^ this part expands to e.g. `m Text`
```

We can have arbitrary type computations, so we could e.g. add another type function to `HasResolver`:

```haskell
class HasResolver (m :: Type -> Type) (a :: Type) where
  type Handler m a
  type NonMonadicHandler a
  type instance NonMonadicHandler a = a
  resolve :: Handler m a -> Maybe (SelectionSetByType Value) -> m (Result Value)
```

and then use that to avoid introducing the extra `m`. This makes it slightly harder to reason about types but practically it might be more intuitive.

**Comment by teh** [comment]
Ah, no, we still need a way of patching through `m`, and also `NonMonadicHandler` doesn't work when we expect `Handler` as input to `resolve`. I'll ponder.

**Comment by teh** [comment]
UX-wise it'd be nicest if we could return maybes directly, e.g. `pure (Just "hello")` but then we can't use `Handler` type expansion because that introduces the extra `m` (and it's needed in the case that we return an object, not just a single `Text` field).

So just to keep composition I don't think we'll get around `Just (pure "hello")`, no matter how ugly.

A 2nd option is trying to peel off the outer `m`, so instead of `pure (Just (pure "hello"))` we'd have just `Just (pure "hello")`, but then we need to figure out a way of running the handler in `m`, i.e. what is `resolve` in the following?:

```haskell
instance forall m hg. (HasResolver m hg, Monad m) => HasResolver m (Maybe hg) where
  type Handler m (Maybe hg) = Maybe (Handler m hg)
  resolve handler selectionSet = ???
```

Slight tangent: We discussed a while back having a `PlainField` type that doesn't run in `m`, maybe that'd be a way of cushioning the API UX issues.

File: CHANGELOG.rst [hunk 1]
```diff
@@ -3,10 +3,16 @@ graphql-api changelog
...
-==================
+=================
+
+
+v0.2.0 (2017-10-12)
+===================
 
 * Make ``Name`` an overloaded string that panics if an invalid name is
   provided.
+* Correctly descent into the type parameter of a `Maybe`. See https://github.com/jml/graphql-api/issues/119.
```
Comment (jml) [inline CHANGELOG.rst]: `descend`

File: CHANGELOG.rst [hunk 2]
```diff
@@ -3,10 +3,16 @@ graphql-api changelog
...
+  This is a backwards-incompatible change.
```
Comment (jml) [inline CHANGELOG.rst]: Can you give an example of what would need to be updated?

File: tests/ResolverTests.hs [hunk 3]
```diff
@@ -37,6 +37,33 @@ tHandler :: Handler TMonad T
 tHandler =
   pure $ (pure 10) :<> (\tArg -> pure tArg) :<> (pure . (*2))
 
+
+-- https://github.com/jml/graphql-api/issues/119
+-- Maybe X didn't descent into its argument. Now it does.
```
Comment (jml) [inline tests/ResolverTests.hs]: `descend`

File: tests/ResolverTests.hs [hunk 4]
```diff
@@ -37,6 +37,33 @@ tHandler :: Handler TMonad T
...
+
+viewFoo :: ServerFoo -> Handler IO Foo
+viewFoo ServerFoo { name=name } = pure $ pure $ name
+
+handler :: Handler IO Query
+handler = pure $ \fooId -> do
+  foo <- lookupFoo fooId
+  -- note that fmap maps over the Maybe, so we still need
+  -- have to wrap the result in a pure.
+  sequence $ fmap (pure . viewFoo) foo
```
Comment (jml) [inline tests/ResolverTests.hs]: A little bit weird that this isn't executed. I guess we're relying on the compiler?

File: tests/ResolverTests.hs (hunk 3 above)
Comment (jml) [reply tests/ResolverTests.hs]: Still needs fixing.

Comment (teh) [reply tests/ResolverTests.hs]: done

File: tests/ResolverTests.hs (hunk 4 above)
Comment (teh) [reply tests/ResolverTests.hs]: fixed

**Pull request merged**